### PR TITLE
Add difficulty adjustment mod

### LINKED ISSUE #3201: Map settings
I think a new UI for configuring maps before playing would be a cool feature to add.

Being able to change settings like:
* Playback rate (more finely controllable than DT/HT)
* AR (instead of EZ/HR which have other effects)
* HP (instead of EZ/HR which have other effects)
* OD (instead of EZ/HR which have other effects)

Changing the settings would make it unranked.

It would practically be a replacement for some mods for practicing or doing cool plays. Also open for more settings suggestions.

It could also be a place where Mods could put in settings (like FL radius, floatiness the new arrange mod which is being discussed in the PRs, hidden fade in time, etc.)

Maybe it would be placed in the beatmap settings thing at the top right corner when you start a map (where you can enable/disable storyboard, change dim, etc.)

I am considering adding it, how would such a feature be received?

### PR DESCRIPTION
**Pull request by LeNitrous**
Adds a new mod "Difficulty Adjust" where the user can override a beatmap's difficulty settings.
Resolves #3201.

Requires https://github.com/ppy/osu/pull/7175 for `IApplicableToBeatmapDifficulty` to apply during song selection.

- [x] Default not updating correctly on sliders (depends on https://github.com/ppy/osu-framework/pull/3123 and subsequent logic changes osu-side). Can probably go ahead with this PR without that being fixed as long as it isn't forgotten.
- [x] Depends on https://github.com/ppy/osu/pull/7303 to show up-to-date statistics in song select details.
- [x] Split implementation into per-ruleset.

### CODE REVIEW

**Comment by peppy** [comment]
can this feasibly just be settings on HR?

**Comment by LeNitrous** [comment]
Sounds fine.

**Comment by peppy** [comment]
I'd try and make the settings on easy / HR (limited in opposite directions), as multipliers.

**Comment by LeNitrous** [comment]
I just realized that HR in the osu! ruleset alters the positions of the hitobjects and what I'm aiming for is to purely change the beatmap settings alone.

**Comment by peppy** [comment]
fair enough

**Comment by peppy** [comment]
Would it not make more sense to draw defaults from the selected beatmap? I'm not 100% sure how most people would use, but it seems like you'd be adjusting from the mapper's settings rather than always using a set combination.

**Comment by LeNitrous** [comment]
It would certainly make more sense with that.

**Comment by peppy** [comment]
Are you going to give that a shot? Should be a pretty simple change.

**Comment by LeNitrous** [comment]
I will. I'm still skimming over the code to know how to set the slider's default values based on the beatmap since `IApplicableToDifficulty` only applies when we move over to the `Player` screen.

**Comment by peppy** [comment]
I see. Changes may need to happen to have this trigger from song select, on beatmap change.

**Comment by peppy** [comment]
- Default not updating correctly on sliders (depends on https://github.com/ppy/osu-framework/pull/3123 and subsequent logic changes osu-side). Can probably go ahead with this PR without that being fixed as long as it isn't forgotten.
- Display values not updating after mod settings change. This needs to be fixed (as a separate PR) before this is merged.

**Comment by peppy** [comment]
@LeNitrous this needs to be split out into per-ruleset implementations. Mania shouldn't have "circle size" adjustments, for instance.

**Comment by LeNitrous** [comment]
I've taken considerations with Taiko and Mania where CS and AR are unused according to the wiki

For Circle Size:
> This option is visible in osu!taiko, but it does not do anything that affects gameplay.

> Key Count, also known as Keys, replaces the Circle Size metric for osu!mania-specific beatmaps.

For Approach Rate:
> This option is visible in osu!taiko and osu!mania, but it does not do anything that affects gameplay. This is mainly because both use a scroll speed, which is based on BPM (beats per minute).

File: osu.Game/Rulesets/Mods/ModDifficultyAdjust.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+        public override ModType Type => ModType.Conversion;
+
+        public override IconUsage Icon => FontAwesome.Solid.Hammer;
+
+        public override double ScoreMultiplier => 1.0;
+
+        public override Type[] IncompatibleMods => new[] { typeof(ModEasy), typeof(ModHardRock) };
+
+        public virtual BindableNumber<float> DrainRate { get; }
```
Comment (peppy) [inline osu.Game/Rulesets/Mods/ModDifficultyAdjust.cs]: `DrainRate` is present in all rulesets, so can be in the base class.

File: osu.Game/Rulesets/Mods/ModDifficultyAdjust.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+        public virtual BindableNumber<float> CircleSize { get; }
```
Comment (peppy) [inline osu.Game/Rulesets/Mods/ModDifficultyAdjust.cs]: Move unused members *out* of the base class.

File: osu.Game/Rulesets/Mods/ModDifficultyAdjust.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        public virtual void ApplyToDifficulty(BeatmapDifficulty difficulty)
+        {
+            if (this.difficulty == null || this.difficulty.ID != difficulty.ID)
+            {
+                this.difficulty = difficulty;
+
+                if (DrainRate != null)
+                    DrainRate.Value = DrainRate.Default = difficulty.DrainRate;
+
+                if (CircleSize != null)
```
Comment (peppy) [inline osu.Game/Rulesets/Mods/ModDifficultyAdjust.cs]: do this in the implementation, not base class.

File: osu.Game/Rulesets/Mods/ModDifficultyAdjust.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+        public override Type[] IncompatibleMods => new[] { typeof(ModEasy), typeof(ModHardRock) };
+
+        public virtual BindableNumber<float> DrainRate { get; }
+
+        public virtual BindableNumber<float> CircleSize { get; }
+
+        public virtual BindableNumber<float> ApproachRate { get; }
+
+        public virtual BindableNumber<float> OverallDifficulty { get; }
```
Comment (peppy) [inline osu.Game/Rulesets/Mods/ModDifficultyAdjust.cs]: same for OD. it's in all four rulesets so didn't need to be moved.

**Review by peppy — CHANGES_REQUESTED** [review]

**Pull request merged**